pkg/controller/cluster/agent/virtual.go

```diff
@@ -0,0 +1,223 @@
+package agent
+
+import (
+	"fmt"
+
+	"github.com/rancher/k3k/pkg/apis/k3k.io/v1alpha1"
+	"github.com/rancher/k3k/pkg/controller"
+	apps "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	VirtualNodeMode      = "virtual"
+	virtualNodeAgentName = "kubelet"
+)
+
+type VirtualAgent struct {
+	cluster   *v1alpha1.Cluster
+	serviceIP string
+}
+
+func NewVirtualAgent(cluster *v1alpha1.Cluster, serviceIP string) Agent {
+	return &VirtualAgent{
+		cluster:   cluster,
+		serviceIP: cluster.Spec.Mode,
+	}
+}
+
+func (v *VirtualAgent) Config() (ctrlruntimeclient.Object, error) {
+	config := virtualAgentData(v.serviceIP, v.cluster.Spec.Token)
+
+	return &v1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      configSecretName(v.cluster.Name),
+			Namespace: v.cluster.Namespace,
+		},
+		Data: map[string][]byte{
+			"config.yaml": []byte(config),
+		},
+	}, nil
+}
+
+func (v *VirtualAgent) Resources() ([]ctrlruntimeclient.Object, error) {
```

Comment:
We can probably update this, the interface, and the other implementations to not return an error unless the other implementations do.